src/main/java/com/thomashan/collection/immutable/ImmutableSetImpl.java

```diff
@@ -0,0 +1,129 @@
+package com.thomashan.collection.immutable;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+/* default */
+@SuppressWarnings("PMD.TooManyMethods")
+class ImmutableSetImpl<E> implements ImmutableSet<E> {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [A class which only has private constructors should be final](https://app.codacy.com/app/thomashan/coup-java/pullRequest?prid=4561385)